### PR TITLE
Packet encoding optimization

### PR DESCRIPTION
**Pull request by thedevop**
As described in #342.

### CODE REVIEW

**Comment by mochi-co** [comment]
I love the idea -how do we feel about the risk of active items in the pool being dropped under high load situations? Is this risk present?

**Comment by thedevop** [comment]
No active items will be in the pool. Currently a buffer is 1st taken out of the pool, if none is available, a new buffer is generated. It will only be put back into the pool when it's no longer needed.

Any items in the pool can be GCed any time, which will have no negative impact.

**Comment by thedevop** [comment]
@werbenhu, I have checked in benchmark, the results are:
```BenchmarkBufferPool
cpu: Intel(R) Core(TM) i7-7920HQ CPU @ 3.10GHz
BenchmarkBufferPool
BenchmarkBufferPool-8                	52733648	        20.26 ns/op	       0 B/op	       0 allocs/op
BenchmarkBufferPoolWithCapLarger
BenchmarkBufferPoolWithCapLarger-8   	56555937	        21.71 ns/op	       0 B/op	       0 allocs/op
BenchmarkBufferPoolWithCapLesser
BenchmarkBufferPoolWithCapLesser-8   	 6784573	       182.1 ns/op	     112 B/op	       2 allocs/op
BenchmarkBufferWithoutPool
BenchmarkBufferWithoutPool-8         	26074543	        49.38 ns/op	      64 B/op	       1 allocs/op
```
Actual performance should be even better, as the bytes.Buffer may grow few times to encode the packet.

I used to use the simplified method as you have provided, which works perfectly fine if all the items are similar in size. It doesn't work as well if items are of different size. See https://golang.org/issue/23199. Although most of the packets without payload should be similar in size, but with the inclusion of metadata (user properties), it may become quite large.

I intentionally for this version avoid the final buffer that includes the payload, we should use a dedicated buffer pool for that, and perhaps let it be user configurable based on their use case.

**Comment by thedevop** [comment]
I benchmarked pk.Properties.Encode using the following:

```go
func BenchmarkEncodeNoPool(b *testing.B) {
	props := propertiesStruct

	for i := 0; i < b.N; i++ {
		buf := bytes.NewBuffer([]byte{})
		props.Encode(Reserved, Mods{AllowResponseInfo: true}, buf, 0)
		_ = buf
	}
}

func BenchmarkEncodePoolBuf(b *testing.B) {
	props := propertiesStruct

	for i := 0; i < b.N; i++ {
		buf := mempool.GetBuffer()
		props.Encode(Reserved, Mods{AllowResponseInfo: true}, buf, 0)
		mempool.PutBuffer(buf)
	}
}
```

The results are:
```
cpu: Intel(R) Core(TM) i7-7920HQ CPU @ 3.10GHz
BenchmarkEncodeNoPool-8    	  664742	      1716 ns/op	     736 B/op	       9 allocs/op
BenchmarkEncodePoolBuf-8   	  748135	      1574 ns/op	     504 B/op	       6 allocs/op
```

If I also replace the buf within Encode to use buffer pool, the results are:
```
cpu: Intel(R) Core(TM) i7-7920HQ CPU @ 3.10GHz
BenchmarkEncodeNoPool-8    	  730120	      1552 ns/op	     288 B/op	       6 allocs/op
BenchmarkEncodePoolBuf-8   	  873458	      1352 ns/op	      56 B/op	       3 allocs/op
```

File: mempool/bufpool.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+// Get a Buffer from the pool.
+func (b *BufferWithCap) Get() *bytes.Buffer {
+	return b.bp.Get()
+}
+
+// Put the Buffer back into the pool if the capacity doesn't exceed the limit. It resets the Buffer
+// for reuse.
+func (b *BufferWithCap) Put(x *bytes.Buffer) {
+	if x.Cap() > b.max {
```
Comment (werbenhu) [inline mempool/bufpool.go:77]: Perhaps adding some comments here would be more beneficial for understanding the code, such as 
```go
// Proper usage of a sync.Pool requires each entry to have approximately
// the same memory cost. To obtain this property when the stored type
// contains a variably-sized buffer, we add a hard limit on the maximum buffer
// to place back in the pool.
//
// See https://golang.org/issue/23199
// Also refer to https://go-review.googlesource.com/c/go/+/136035
if x.Cap() > b.max {
```

**Comment by werbenhu** [comment]
@thedevop @mochi-co Can we consider retaining only the usage of BufferWithCap and eliminating the approach of using Buffer Poll without a cap? 

See https://go-review.googlesource.com/c/go/+/136116 and https://go-review.googlesource.com/c/go/+/136035 for reference.

**Comment by werbenhu** [comment]
@thedevop perhaps we can learn from [encoding/json: fix usage of sync.Pool](https://go-review.googlesource.com/c/go/+/136035), where we can see that the standard library's json.Marshal also faces a similar issue to what we are currently encountering. However, at least for now, it appears that the json.Marshal still directly employs the sync.Pool without a cap. Personally, I suggest we can temporarily postpone the use of BufferWithCap. @mochi-co  what are your thoughts on this issue?

**Comment by thedevop** [comment]
@werbenhu, I'm ok we use the pool without cap, as most of the packet encoding without payload should vary too much on size. I updated the default pool to not have a cap.

**Review by werbenhu — APPROVED** [review]

**Comment by mochi-co** [comment]
Sorry my life has been crazy this year. Just catching up on all this good work you've done.

The pool benchmarks are promising, and at scale this will make a significant difference. @thedevop your point about differing payload sizes is poignant and well considered.

I agree with @werbenhu that we should first attempt to use the pool without a cap to reduce the complexity we're adding here. However, we should monitor the change and add the cap if it does become necessary.

As for me, I think this is a great contribution, and it's definitely getting merged.

**Review by mochi-co — APPROVED** [review]

**Pull request merged**